File: backend/app/utils/config_summary.py
"""Human-readable configuration lines for orders, admin UI, and emails."""
from __future__ import annotations

import re
from typing import Any, Dict, List, Optional

CONFIG_LABELS = {
    "engine": "Engine",
    "chassisType": "Chassis",
    "finish": "Finish",
    "handThrottle": "Hand throttle",
    "propeller": "Propeller",
    "color": "Color",
    "colorId": "Color",
    "chassisColor": "Chassis color",
    "accentColor": "Accent color",
    "peripheralColor": "Peripheral color",
    "firstName": "Guest first name",
    "lastName": "Guest last name",
    "phone": "Phone",
    "age": "Age",
    "locationId": "Location",
    "locationName": "Location",
    "duration": "Duration",
}

UPGRADE_LABELS = {
    "sun-roof-netting": "Sun-roof netting",
    "cruise-control": "Cruise control",
    "camel-back": "Camel back",
    "instrument-kit": "Instrument kit",
    "electrical-kit": "Electrical kit",
    "rear-mirror": "Rear mirror",
    "front-brake": "Front brake",
    "front-bar-protection": "Padded roll bar",
    "cockpit-liner": "Cockpit liner",
    "parachute-container": "Parachute container",
    "lateral-bag": "Side explorer cases",
    "lateral-bag-explorer": "Lateral bag explorer",
    "bottom-explorer-bag": "Bottom explorer bag",
    "reserve-chute": "Reserve parachute",
    "auxiliary-lights": "Auxiliary lights",
    "carabiners": "Carabiners",
    "propeller-guard": "Propeller guard",
    "rock-guard": "Rock guard",
    "fuel-gauge-vanguard": "Fuel gauge",
}


def extract_option_id(value: Any) -> Optional[str]:
    if value is None:
        return None
    if isinstance(value, str):
        cleaned = value.strip()
        return cleaned or None
    if isinstance(value, (int, float, bool)):
        return str(value)
    if isinstance(value, dict):
        for key in ("id", "slug", "name", "displayName", "hex"):
            raw = value.get(key)
            if raw is not None and str(raw).strip():
                return str(raw).strip()
    return None


def _fmt_slug(value: Any) -> str:
    raw = extract_option_id(value) or str(value or "").strip()
    if not raw:
        return ""
    if raw in UPGRADE_LABELS:
        return UPGRADE_LABELS[raw]
    cleaned = re.sub(r"^(vanguard|nomadic|acc|part)-", "", raw, flags=re.I)
    cleaned = cleaned.replace("-", " ").replace("_", " ").strip()
    if not cleaned:
        return raw
    return cleaned.title()


def format_config_lines(config: Optional[Dict[str, Any]]) -> List[Dict[str, str]]:
    if not config or not isinstance(config, dict):
        return []

    lines: List[Dict[str, str]] = []

    if config.get("bookingType") == "tourist-flight":
        guest = " ".join(
            part for part in (config.get("firstName"), config.get("lastName")) if part
        ).strip()
        if guest:
            lines.append({"label": "Guest", "value": guest})
        for key in ("phone", "age", "locationName", "locationId", "duration"):
            val = config.get(key)
            if val:
                label = CONFIG_LABELS.get(key, key.replace("_", " ").title())
                lines.append({"label": label, "value": _fmt_slug(val) if key.endswith("Id") else str(val)})
        return lines

    for key in (
        "engine", "chassisType", "finish", "handThrottle", "propeller",
        "color", "colorId", "chassisColor", "accentColor", "peripheralColor",
    ):
        val = config.get(key)
        if val:
            lines.append({"label": CONFIG_LABELS.get(key, key), "value": _fmt_slug(val)})

    upgrades = config.get("upgrades")
    if isinstance(upgrades, list) and upgrades:
        labels = [_fmt_slug(item) for item in upgrades if extract_option_id(item) or item]
        labels = [label for label in labels if label]
        if labels:
            lines.append({"label": "Accessories", "value": ", ".join(labels)})

    return lines


def format_config_text(config: Optional[Dict[str, Any]]) -> str:
    return "; ".join(
        f"{line['label']}: {line['value']}" for line in format_config_lines(config)
    )

File: frontend/lib/configSummary.js
/** Human-readable configuration lines — shared by cart, orders, and admin. */

const UPGRADE_LABELS = {
  'sun-roof-netting': 'Sun-roof netting',
  'cruise-control': 'Cruise control',
  'camel-back': 'Camel back',
  'instrument-kit': 'Instrument kit',
  'electrical-kit': 'Electrical kit',
  'rear-mirror': 'Rear mirror',
  'front-brake': 'Front brake',
  'front-bar-protection': 'Padded roll bar',
  'cockpit-liner': 'Cockpit liner',
  'parachute-container': 'Parachute container',
  'lateral-bag': 'Side explorer cases',
  'lateral-bag-explorer': 'Lateral bag explorer',
  'bottom-explorer-bag': 'Bottom explorer bag',
  'reserve-chute': 'Reserve parachute',
  'auxiliary-lights': 'Auxiliary lights',
  carabiners: 'Carabiners',
  'propeller-guard': 'Propeller guard',
  'rock-guard': 'Rock guard',
  'fuel-gauge-vanguard': 'Fuel gauge',
}

export function extractOptionId(value) {
  if (value == null) return null
  if (typeof value === 'string') return value.trim() || null
  if (typeof value === 'number' || typeof value === 'boolean') return String(value)
  if (typeof value === 'object') {
    return value.id || value.slug || value.name || value.displayName || value.hex || null
  }
  return null
}

export function fmtOption(value) {
  const raw = extractOptionId(value) || String(value ?? '').trim()
  if (!raw) return ''
  if (UPGRADE_LABELS[raw]) return UPGRADE_LABELS[raw]
  const stripped = raw.replace(/^(vanguard|nomadic|acc|part)-/i, '')
  return stripped.replace(/-/g, ' ').replace(/\b\w/g, (c) => c.toUpperCase())
}

/** Normalize configurador payload before POST /carrito — store slugs, not full objects. */
export function normalizeConfigForApi(config = {}) {
  const pick = (value) => extractOptionId(value)
  const pickUpgrade = (item) => {
    const id = extractOptionId(item)
    return id ? String(id) : null
  }

  const normalized = {
    engine: pick(config.engine),
    chassisType: pick(config.chassisType),
    finish: pick(config.finish),
    handThrottle: pick(config.handThrottle),
    propeller: pick(config.propeller),
    color: pick(config.color),
    colorId: pick(config.colorId),
    upgrades: (config.upgrades || []).map(pickUpgrade).filter(Boolean),
    bookingType: config.bookingType,
    firstName: config.firstName,
    lastName: config.lastName,
    phone: config.phone,
    age: config.age,
    locationId: config.locationId,
    locationName: config.locationName,
    duration: config.duration,
    termsAccepted: config.termsAccepted,
  }

  if (config.chassisColor) {
    normalized.chassisColor = typeof config.chassisColor === 'object'
      ? [config.chassisColor.name, config.chassisColor.hex].filter(Boolean).join(', ')
      : String(config.chassisColor)
  }
  if (config.accentColor) {
    normalized.accentColor = typeof config.accentColor === 'object'
      ? [config.accentColor.name, config.accentColor.hex].filter(Boolean).join(', ')
      : String(config.accentColor)
  }

  return Object.fromEntries(
    Object.entries(normalized).filter(([, value]) => value != null && value !== ''),
  )
}

export function formatConfigSummary(config) {
  if (!config || typeof config !== 'object') return []

  if (Array.isArray(config.config_summary) && config.config_summary.length) {
    return config.config_summary
  }

  const source = config.configuracion && typeof config.configuracion === 'object'
    ? config.configuracion
    : config

  const lines = []

  if (source.bookingType === 'tourist-flight') {
    if (source.firstName || source.lastName) {
      lines.push({
        label: 'Guest',
        value: [source.firstName, source.lastName].filter(Boolean).join(' '),
      })
    }
    if (source.phone) lines.push({ label: 'Phone', value: source.phone })
    if (source.age) lines.push({ label: 'Age', value: String(source.age) })
    if (source.locationName || source.locationId) {
      lines.push({ label: 'Location', value: source.locationName || fmtOption(source.locationId) })
    }
    if (source.duration) lines.push({ label: 'Duration', value: source.duration })
    return lines
  }

  if (source.engine) lines.push({ label: 'Engine', value: fmtOption(source.engine) })
  if (source.chassisType) lines.push({ label: 'Chassis', value: fmtOption(source.chassisType) })
  if (source.finish) lines.push({ label: 'Finish', value: fmtOption(source.finish) })
  if (source.handThrottle) lines.push({ label: 'Hand throttle', value: fmtOption(source.handThrottle) })
  if (source.propeller) lines.push({ label: 'Propeller', value: fmtOption(source.propeller) })
  if (source.color || source.colorId) {
    lines.push({ label: 'Color', value: fmtOption(source.color || source.colorId) })
  }
  if (source.chassisColor) lines.push({ label: 'Chassis color', value: fmtOption(source.chassisColor) })
  if (source.accentColor) lines.push({ label: 'Accent color', value: fmtOption(source.accentColor) })

  if (Array.isArray(source.upgrades) && source.upgrades.length) {
    lines.push({
      label: 'Accessories',
      value: source.upgrades.map((u) => fmtOption(u)).filter(Boolean).join(', '),
    })
  }

  return lines
}
